### PR TITLE
feat: quota improvement for bucket migration

### PR DESCRIPTION
**Pull request by jingjunLi**
### Description

1. During the bucket migration approval phase, add a quota check. If there is insufficient quota, reject the migration request outright.
2. When generating the bucket migration execution plan, pre-deduct the quota required for bucket migration from the source SP (Storage Provider).
3. After a successful bucket migration, read the quota from the source SP and allocate it to the destination SP node.
4. In case of a failed bucket migration, refund the pre-deducted quota to the user. The specific deduction of quota is based on the size of the migrated objects.
image

### Rationale

* base/gfspclient/gater.go interface
``` go
QueryLatestBucketQuota(ctx context.Context, endpoint string, bucketID uint64) (types.GfSpQuotaInfo, error)
PreMigrateBucket(ctx context.Context, srcSPEndpoint string, bucketID uint64) (types.GfSpQuotaInfo, error)
PostMigrateBucket(ctx context.Context, srcSPEndpoint string, postMsg *types.GfSpPostMigrateBucket) (types.GfSpQuotaInfo, error)
QuerySPHasEnoughQuotaForMigrateBucket(ctx context.Context, srcSPEndpoint string, bucketID uint64) error
```
* modular/gater/migrate_handler.go
``` go
getLatestBucketQuotaHandler
preMigrateBucketHandler
postMigrateBucketHandler
hasEnoughQuotaForMigrateBucketHandler
```

add a new interface for SPDB
``` go
GetLatestBucketTraffic
UpdateBucketTraffic
```

### Example

NA

### Changes

Notable changes: 
* NA
* ...

### CODE REVIEW

File: modular/approver/approve_task.go [hunk 1]
```diff
@@ -149,6 +149,12 @@ func (a *ApprovalModular) HandleMigrateBucketApprovalTask(ctx context.Context, t
 		return true, nil
 	}
 
+	// check src sp has enough quota
+	succeed, err := a.migrateBucketQuotaCheck(ctx, task)
```
Comment (sysvm) [inline modular/approver/approve_task.go:153]: fix `succeed` as `success`, `succeed` is a verb.

File: modular/approver/approve_task_test.go [hunk 2]
```diff
@@ -4,15 +4,20 @@ import (
 	"context"
 	"testing"
 
+	"cosmossdk.io/math"
+
```
Comment (sysvm) [inline modular/approver/approve_task_test.go:8]: remove this meaningless blank line.

File: modular/blocksyncer/blocksyncer_options.go [hunk 3]
```diff
@@ -505,3 +511,46 @@ func mustGetLatestHeight(ctx *parser.Context) uint64 {
...
+		if len(buckets) == 0 {
+			continue
+		}
+		err = db.Cast(b.parserCtx.Database).BatchUpdateBucketSize(context.Background(), buckets)
+		if err != nil {
+			log.Errorw("failed to update size", "error", err)
+			return err
+		}
+	}
+	log.Infof("sync bucket size success")
```
Comment (sysvm) [inline modular/blocksyncer/blocksyncer_options.go:554]: Please use `log.Info` not `log.Infof`.

File: modular/gater/migrate_handler.go [hunk 4]
```diff
@@ -299,3 +301,263 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
...
+		if migratedBytes > bucketSize {
+			log.CtxErrorw(reqCtx.Context(), "failed to recoup extra quota to user", "error", err)
+			return
+		}
+		quotaUpdateErr := g.baseApp.GfSpClient().RecoupQuota(reqCtx.Context(), postMigrateBucket.GetBucketId(), extraQuota, quota.GetMonth())
+		// no need to return the db error to user
+		if quotaUpdateErr != nil {
+			log.CtxErrorw(reqCtx.Context(), "failed to recoup extra quota to user", "error", err)
+		}
+		log.CtxDebugw(reqCtx.Context(), "//"+
```
Comment (sysvm) [inline modular/gater/migrate_handler.go:497]: What does `//` mean?

File: store/bsdb/bucket.go [hunk 5]
```diff
@@ -4,6 +4,8 @@ import (
 	"errors"
 	"time"
 
+	"github.com/shopspring/decimal"
+
```
Comment (sysvm) [inline store/bsdb/bucket.go:8]: Delete this meaningless blank line.

File: modular/gater/migrate_handler.go [hunk 6]
```diff
@@ -299,3 +301,261 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
...
+		if err != nil {
+			reqCtx.SetError(err)
+			reqCtx.SetHTTPCode(int(gfsperrors.MakeGfSpError(err).GetHttpStatusCode()))
+			MakeErrorResponse(w, gfsperrors.MakeGfSpError(err))
+		} else {
+			reqCtx.SetHTTPCode(http.StatusOK)
+		}
+		log.CtxDebugw(reqCtx.Context(), reqCtx.String())
+	}()
+
```
Comment (flywukong) [inline modular/gater/migrate_handler.go:377]: Lack of authentication

File: modular/gater/migrate_handler.go [hunk 7]
```diff
@@ -299,3 +301,261 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
...
```
Comment (flywukong) [inline modular/gater/migrate_handler.go:446]: Lack of authentication

File: modular/gater/migrate_handler.go [hunk 8]
```diff
@@ -299,3 +301,261 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
...
+	}
+
+	w.Header().Set(ContentTypeHeader, ContentTypeXMLHeaderValue)
+
+	log.CtxInfow(reqCtx.Context(), "succeed to post bucket migrate", "bucket_id",
+		bucketID, "quota", quota, "postMigrateBucket", postMigrateBucket, "postMsg", postMsg)
+}
+
+// hasEnoughQuotaForMigrateBucketHandler notifying the source sp about the completion of migration bucket
+func (g *GateModular) hasEnoughQuotaForMigrateBucketHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (flywukong) [inline modular/gater/migrate_handler.go:506]: function name meaning is not same with the comment..  name is strange

File: modular/gater/migrate_handler.go [hunk 9]
```diff
@@ -299,3 +301,261 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
...
+			reqCtx.SetError(err)
+			reqCtx.SetHTTPCode(int(gfsperrors.MakeGfSpError(err).GetHttpStatusCode()))
+			MakeErrorResponse(w, gfsperrors.MakeGfSpError(err))
+		} else {
+			reqCtx.SetHTTPCode(http.StatusOK)
+		}
+		log.CtxDebugw(reqCtx.Context(), reqCtx.String())
+	}()
+
+	reqCtx, _ = NewRequestContext(r, g)
```
Comment (flywukong) [inline modular/gater/migrate_handler.go:526]: Lack of authentication

File: modular/gater/migrate_handler.go [hunk 10]
```diff
@@ -299,3 +301,261 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
...
+		if err != nil {
+			return
+		}
+		migratedBytes := postMigrateBucket.GetMigratedBytesSize()
+		extraQuota := bucketSize - migratedBytes
+		if migratedBytes > bucketSize {
+			log.CtxErrorw(reqCtx.Context(), "failed to recoup extra quota to user", "error", err)
+			return
+		}
+		quotaUpdateErr := g.baseApp.GfSpClient().RecoupQuota(reqCtx.Context(), postMigrateBucket.GetBucketId(), extraQuota, quota.GetMonth())
```
Comment (flywukong) [inline modular/gater/migrate_handler.go:491]: RecoupQuota should be called in source SP , not dest SP, check if it is right here

File: modular/gater/query_util.go [hunk 11]
```diff
@@ -112,6 +114,23 @@ func (g *GateModular) checkSPAndBucketStatus(ctx context.Context, bucketName str
 	return nil
 }
 
+// checkSPAndBucketStatus check sp and bucket is in right status
+func (g *GateModular) getBucketTotalSize(ctx context.Context, bucketID uint64) (uint64, error) {
```
Comment (flywukong) [inline modular/gater/query_util.go:118]: the bucket size is allows changing.  ensure it has been called after migration start

File: base/gfspclient/gater.go [hunk 12]
```diff
@@ -208,6 +224,137 @@ func (s *GfSpClient) NotifyDestSPMigrateSwapOut(ctx context.Context, destEndpoin
...
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("failed to query latest bucket quota, bucket(%d), status_code(%d), endpoint(%s)", bucketID, resp.StatusCode, srcSPEndpoint)
+	}
+
+	signedMsg := resp.Header.Get(GnfdSignedApprovalMsgHeader)
+
+	if signedMsg == "true" {
```
Comment (flywukong) [inline base/gfspclient/gater.go:351]: it is not safe if just a "true",  may be use the private key to sign the approval,,

File: base/gfspapp/sign_server.go [hunk 13]
```diff
@@ -220,6 +220,12 @@ func (g *GfSpBaseApp) GfSpSign(ctx context.Context, req *gfspserver.GfSpSignRequ
 		if err != nil {
 			log.CtxErrorw(ctx, "failed to sign migrate gvg task", "error", err)
 		}
+	case *gfspserver.GfSpSignRequest_GfspMigrateBucketStatus:
+		ctx = log.WithValue(ctx, log.CtxKeyTask, t.GfspMigrateBucketStatus.Key().String())
+		signature, err = g.signer.SignMigrateBucket(ctx, t.GfspMigrateBucketStatus)
+		if err != nil {
+			log.CtxErrorw(ctx, "failed to sign migrate bucket task", "error", err)
```
Comment (flywukong) [inline base/gfspapp/sign_server.go:227]: log should print bucket name

File: base/gfspclient/downloader.go [hunk 14]
```diff
@@ -82,6 +82,31 @@ func (s *GfSpClient) RecoupQuota(ctx context.Context, bucketID, extraQuota uint6
...
+	defer conn.Close()
+	req := &gfspserver.GfSpDeductQuotaForBucketMigrateRequest{
+		BucketId:    bucketID,
+		DeductQuota: deductQuota,
+		YearMonth:   yearMonth,
+	}
+
+	resp, err := gfspserver.NewGfSpDownloadServiceClient(conn).GfSpDeductQuotaForBucketMigrate(ctx, req)
+	if err != nil {
+		log.CtxErrorw(ctx, "client failed to deduct the quota for bucket migrate", "req", req, "error", err)
```
Comment (flywukong) [inline base/gfspclient/downloader.go:100]: log should print bucket name

File: base/gfspclient/gater.go [hunk 15]
```diff
@@ -208,6 +219,154 @@ func (s *GfSpClient) NotifyDestSPMigrateSwapOut(ctx context.Context, destEndpoin
...
+	}
+	msg, err := json.Marshal(preMsg)
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to pre migrate bucket due to marshal error", "error", err)
+		return gfsptask.GfSpBucketQuotaInfo{}, err
+	}
+	req.Header.Add(GnfdMigrateBucketMsgHeader, hex.EncodeToString(msg))
+	resp, err := s.HTTPClient(ctx).Do(req)
+	if err != nil {
+		log.Errorw("failed to pre migrate bucket msg", "bucket_msg", preMsg, "error", err)
```
Comment (flywukong) [inline base/gfspclient/gater.go:274]: log should print bucket name

File: modular/gater/migrate_handler.go [hunk 16]
```diff
@@ -101,6 +103,65 @@ func (g *GateModular) checkMigratePieceAuth(reqCtx *RequestContext, migrateGVGHe
...
+		return allow, migrateBucketStatus, err
+	}
+
+	effect, err := g.baseApp.GfSpClient().VerifyMigrateGVGPermission(reqCtx.Context(), migrateBucketStatus.GetBucketId(), 0 /*not used*/, sp.GetId())
+	if effect == nil || err != nil {
+		// metadate serive support cancel & completion migrate bucket
+		if strings.Contains(err.Error(), "the bucket is not in migration status") {
+			return true, migrateBucketStatus, nil
+		}
+		log.Errorw("failed to verify migrate gvg permission", "bucket_migration", migrateBucketStatus,
```
Comment (flywukong) [inline modular/gater/migrate_handler.go:120]: log should print bucket name

File: modular/gater/migrate_handler.go [hunk 17]
```diff
@@ -300,3 +361,287 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
...
+		log.CtxDebugw(reqCtx.Context(), reqCtx.String())
+	}()
+
+	reqCtx, _ = NewRequestContext(r, g)
+	if allowMigrate, migrateBucketStatus, _, err = g.verifySignatureAndSP(reqCtx, r.Header.Get(GnfdMigrateBucketMsgHeader)); err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to check migrate bucket auth", "migrate_bucket", migrateBucketStatus, "error", err)
+		return
+	}
+	if !allowMigrate {
+		log.CtxErrorw(reqCtx.Context(), "no permission to migrate piece", "migrate_bucket", migrateBucketStatus)
```
Comment (flywukong) [inline modular/gater/migrate_handler.go:393]: log should print bucket name

File: modular/gater/migrate_handler.go [hunk 18]
```diff
@@ -101,6 +103,67 @@ func (g *GateModular) checkMigratePieceAuth(reqCtx *RequestContext, migrateGVGHe
...
+// checkMigrateBucketQuotaAuth parse migrateBucketMsgHeader to GfSpBucketMigrationInfo and check if request has permission
+// only used by preMigrateBucketHandler and postMigrateBucketHandler
+func (g *GateModular) checkMigrateBucketQuotaAuth(reqCtx *RequestContext, migrateBucketMsgHeader string) (bool, *gfsptask.GfSpBucketMigrationInfo, error) {
+	allow, bucketMigrationInfo, sp, err := g.verifySignatureAndSP(reqCtx, migrateBucketMsgHeader)
+	if !allow {
+		return allow, bucketMigrationInfo, err
+	}
+
+	effect, err := g.baseApp.GfSpClient().VerifyMigrateGVGPermission(reqCtx.Context(), bucketMigrationInfo.GetBucketId(), 0 /*not used*/, sp.GetId())
+	if effect == nil || err != nil {
```
Comment (alexgao001) [inline modular/gater/migrate_handler.go:115]: dont think need to put` effect == nil || err != nil` together and do ` if err != nil` within the if branch. Can check 1 by 1 instead.

File: modular/gater/migrate_handler.go [hunk 19]
```diff
@@ -101,6 +103,67 @@ func (g *GateModular) checkMigratePieceAuth(reqCtx *RequestContext, migrateGVGHe
...
+			// metadate serive support cancel & completion migrate bucket
+			if strings.Contains(err.Error(), "the bucket is not in migration status") {
+				return true, bucketMigrationInfo, nil
+			}
+		}
+		log.Errorw("failed to verify migrate gvg permission", "bucket_migration_info", bucketMigrationInfo,
+			"dest_sp", sp.GetId(), "effect", effect, "error", err)
+		return false, bucketMigrationInfo, err
+	}
+	if *effect == permissiontypes.EFFECT_ALLOW {
```
Comment (alexgao001) [inline modular/gater/migrate_handler.go:126]: return `*effect == permissiontypes.EFFECT_ALLOW, bucketMigrationInfo, nil`

File: modular/gater/migrate_handler.go [hunk 20]
```diff
@@ -300,3 +363,287 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
...
+	quota, err := g.baseApp.GfSpClient().GetLatestBucketReadQuota(
+		reqCtx.Context(), bucketID)
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to get bucket read quota", "bucket_id",
+			bucketID, "error", err)
+		return
+	}
+
+	// reduce quota, sqldb
+	err = g.baseApp.GfSpClient().DeductQuotaForBucketMigrate(
```
Comment (alexgao001) [inline modular/gater/migrate_handler.go:475]: If the dest SP retry calling `preMigrateBucket`, would it deduct multiple times?

File: util/virtual_group.go [hunk 21]
```diff
@@ -86,3 +89,22 @@ func GetBucketPrimarySPID(ctx context.Context, chainClient consensus.Consensus,
 	}
 	return resp.GetPrimarySpId(), nil
 }
+
+// GetBucketPrimarySPInfoByBucketID return bucket sp info by vgf id
+func GetBucketPrimarySPInfoByBucketID(ctx context.Context, chainClient consensus.Consensus, bucketID uint64) (*sptypes.StorageProvider, error) {
```
[the remark this answers is not in the document]
Comment (alexgao001) [reply util/virtual_group.go:94]: `chainClient.QuerySPByID(ctx, globalVGF.GetId())` actually should takes a SP id not VGF ID... `GfSpGetBucketMeta` returns the PrimarySpId, then you can use it to retrieve the SPInfo by Meta ApI as well.

**Review by alexgao001 — APPROVED** [review]

File: base/gfspclient/signer.go [hunk 22]
```diff
@@ -500,3 +500,25 @@ func (s *GfSpClient) SignMigrateGVG(ctx context.Context, task *gfsptask.GfSpMigr
...
+		return nil, ErrRPCUnknownWithDetail("client failed to connect to signer, error: " + err.Error())
+	}
+	req := &gfspserver.GfSpSignRequest{
+		Request: &gfspserver.GfSpSignRequest_GfspBucketMigrateInfo{
+			GfspBucketMigrateInfo: task,
+		},
+	}
+	resp, err := gfspserver.NewGfSpSignServiceClient(conn).GfSpSign(ctx, req)
+	if err != nil {
+		log.CtxErrorw(ctx, "client failed to sign migrate gvg", "migrate_gvg", task, "error", err)
```
Comment (will-2012) [inline base/gfspclient/signer.go:517]: Refine this log.

File: go.mod [hunk 23]
```diff
@@ -302,7 +303,7 @@ replace (
 	github.com/confio/ics23/go => github.com/cosmos/cosmos-sdk/ics23/go v0.8.0
 	github.com/cosmos/cosmos-sdk => github.com/bnb-chain/greenfield-cosmos-sdk v0.2.6
 	github.com/cosmos/iavl => github.com/bnb-chain/greenfield-iavl v0.20.1
-	github.com/forbole/juno/v4 => github.com/bnb-chain/juno/v4 v4.0.0-20231023111452-7cd108f48b21
+	github.com/forbole/juno/v4 => github.com/bnb-chain/juno/v4 v4.0.0-20230921022336-06a9e87ef18f
```
Comment (will-2012) [inline go.mod:306]: Please check this dependency. 1023->0921?

File: modular/executor/migrate_task.go [hunk 24]
```diff
@@ -183,6 +184,37 @@ func (e *ExecuteModular) doObjectMigration(ctx context.Context, gvgTask coretask
...
 }
 
+func (e *ExecuteModular) doObjectMigrationRetry(ctx context.Context, gvgTask coretask.MigrateGVGTask, bucketID uint64, object *metadatatypes.ObjectDetails) error {
+	var (
+		srcGvgID = gvgTask.GetSrcGvg().GetId()
+		err      error
+	)
+	for retry := 0; retry < e.maxObjectMigrationRetry; retry++ {
+		// when cancel migrate bucket, the dest sp event may be slower than src sp, so we retry this migration
+		err = e.doObjectMigration(ctx, gvgTask, bucketID, object)
```
Comment (will-2012) [inline modular/executor/migrate_task.go:194]: if err = xx; err != {
    // check and retry
}

File: modular/gater/migrate_handler.go [hunk 25]
```diff
@@ -183,7 +242,7 @@ func (g *GateModular) migratePieceHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	pieceTask := &gfsptask.GfSpDownloadPieceTask{}
-	pieceTask.InitDownloadPieceTask(chainObjectInfo, bucketInfo, params, coretask.DefaultSmallerPriority, migratePiece.GetIsBucketMigrate(), bucketInfo.Owner,
+	pieceTask.InitDownloadPieceTask(chainObjectInfo, bucketInfo, params, coretask.DefaultSmallerPriority, false, bucketInfo.Owner,
```
Comment (will-2012) [inline modular/gater/migrate_handler.go:245]: Why adjust this code?

File: modular/gater/router_test.go [hunk 26]
```diff
@@ -653,6 +653,14 @@ func TestRouters(t *testing.T) {
...
 			wantedRouterName: verifyMigrateGVGPermissionRouterName,
 		},
+		{
+			name:             "Get bucket total object size",
+			router:           gwRouter,
+			method:           http.MethodGet,
+			url:              fmt.Sprintf("%s%s/?%s&%s", scheme, testDomain, GetBucketSizeQuery, BucketIDQuery),
+			shouldMatch:      true,
+			wantedRouterName: getBucketSizeRouterName,
+		},
```
Comment (will-2012) [inline modular/gater/router_test.go:663]: Add more ut.

File: modular/signer/signer.go [hunk 27]
```diff
@@ -170,6 +170,15 @@ func (s *SignModular) SignMigrateGVG(ctx context.Context, mp *gfsptask.GfSpMigra
 	return sig, nil
 }
 
+func (s *SignModular) SignMigrateBucket(ctx context.Context, mp *gfsptask.GfSpBucketMigrationInfo) ([]byte, error) {
+	sig, err := s.client.Sign(SignOperator, mp.GetSignBytes())
+	if err != nil {
+		log.Errorw("failed to sign migrate gvg", "error", err)
```
Comment (will-2012) [inline modular/signer/signer.go:176]: Refine this log.

File: store/sqldb/traffic.go [hunk 28]
```diff
@@ -374,6 +374,83 @@ func (s *SpDBImpl) UpdateExtraQuota(bucketID, extraQuota uint64, yearMonth strin
...
+		BucketName:            update.BucketName,
+		ReadConsumedSize:      update.ReadConsumedSize,
+		ChargedQuotaSize:      update.ChargedQuotaSize,
+		ModifiedTime:          time.Now(),
+	}
+
+	if needInsert {
+		result = s.db.Create(updateRecord)
+		if result.Error != nil || result.RowsAffected != 1 {
+			return fmt.Errorf("failed to insert record in subscribe progress table: %s", result.Error)
```
Comment (will-2012) [inline store/sqldb/traffic.go:442]: Refine this log.

File: store/sqldb/traffic.go [hunk 29]
```diff
@@ -374,6 +374,83 @@ func (s *SpDBImpl) UpdateExtraQuota(bucketID, extraQuota uint64, yearMonth strin
...
+		}
+	} else { // update
+		result = s.db.Model(&BucketTrafficTable{}).
+			Where("bucket_id = ? and month = ?", bucketID, update.YearMonth).Updates(updateRecord)
+		if result.Error != nil {
+			return fmt.Errorf("failed to update record in subscribe progress table: %s", result.Error)
```
Comment (will-2012) [inline store/sqldb/traffic.go:448]: Ditto.

**Review by leowkong — APPROVED** [review]

**Review by flywukong — APPROVED** [review]

**Review by constwz — APPROVED** [review]
LGTM

**Pull request merged**